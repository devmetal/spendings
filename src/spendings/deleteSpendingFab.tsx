import React, { FC, useState } from 'react';
import {
  CircularProgress,
  createStyles,
  Fab,
  makeStyles,
  Theme,
} from '@material-ui/core';
import DeleteIcon from '@material-ui/icons/Delete';
import { animated, useSpring } from 'react-spring';
import { red } from '@material-ui/core/colors';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    wrapper: {
      position: 'relative',
      margin: theme.spacing(1),
      zIndex: 1,
    },
    fabProgress: {
      color: red[600],
      position: 'absolute',
      top: -40,
      left: 20,
      zIndex: 1,
    },
  })
);

export interface DeleteSpendingFabProps {
  onDelete: () => void;
}

const AnimatedProgress = animated(
  ({ step, className }: { step: number; className: string }) => (
    <CircularProgress
      size={40}
      variant="determinate"
      className={className}
      value={step}
    />
  )
);

export const DeleteSpendingFab: FC<DeleteSpendingFabProps> = ({ onDelete }) => {
  const classes = useStyles();
  const [down, setIsDown] = useState(false);

  const { step } = useSpring({
    config: { duration: 1500, precision: 0.1 },
    step: down ? 5 : 1,
    immediate: !down,
    onRest: () => {
      if (down) {
        onDelete();
      }
    },
  });

  return (
    <div
      className={classes.wrapper}
      onTouchStart={(e) => {
        e.stopPropagation();
        setIsDown(true);
      }}
      onTouchEnd={(e) => {
        e.stopPropagation();
        setIsDown(false);
      }}
      onMouseDown={(e) => {
        e.stopPropagation();
        setIsDown(true);
      }}
      onMouseUp={(e) => {
        e.stopPropagation();
        setIsDown(false);
      }}
      onMouseOut={() => {
        setIsDown(false);
      }}
    >
      <Fab
        size="small"
        aria-label="delete"
        role="delete-item"
        color="secondary"
      >
        <DeleteIcon />
      </Fab>
      <AnimatedProgress
        className={classes.fabProgress}
        step={step
          .to((s) => Math.floor(s))
          .to({
            range: [1, 2, 3, 4, 5],
            output: [0, 25, 50, 75, 100],
          })}
      />
    </div>
  );
};
