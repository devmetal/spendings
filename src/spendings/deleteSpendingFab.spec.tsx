import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { DeleteSpendingFab } from './deleteSpendingFab';

const delay = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

test('Sanity snapshot test', () => {
  const { container } = render(<DeleteSpendingFab onDelete={jest.fn()} />);
  expect(container).toMatchSnapshot();
});

test('Holding down 1500ms fire the onDelete event', async () => {
  const onDelete = jest.fn();

  render(<DeleteSpendingFab onDelete={onDelete} />);

  const evt = document.createEvent('MouseEvents');
  evt.initEvent('mousedown', true, true);

  fireEvent(screen.getByRole('delete-item'), evt);
  await delay(1600);
  expect(onDelete).toBeCalled();
});

test('Holding down and release have to reset the delete counter', async () => {
  const onDelete = jest.fn();

  render(<DeleteSpendingFab onDelete={onDelete} />);

  const down = document.createEvent('MouseEvents');
  const up = document.createEvent('MouseEvents');
  down.initEvent('mousedown', true, true);
  up.initEvent('mouseup', true, true);

  const button = screen.getByRole('delete-item');
  fireEvent(button, down);
  await delay(1000);

  fireEvent(button, up);
  expect(onDelete).not.toBeCalled();

  fireEvent(button, down);
  await delay(1600);
  expect(onDelete).toBeCalledTimes(1);
});
