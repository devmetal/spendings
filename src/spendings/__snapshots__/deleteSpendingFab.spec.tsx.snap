// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sanity snapshot test 1`] = `
<div>
  <div
    class="makeStyles-wrapper-1"
  >
    <button
      aria-label="delete"
      class="MuiButtonBase-root MuiFab-root MuiFab-sizeSmall MuiFab-secondary"
      role="delete-item"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiFab-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      aria-valuenow="0"
      class="MuiCircularProgress-root makeStyles-fabProgress-2 MuiCircularProgress-colorPrimary MuiCircularProgress-determinate"
      role="progressbar"
      style="width: 40px; height: 40px; transform: rotate(-90deg);"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleDeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
          style="stroke-dasharray: 126.920; stroke-dashoffset: 126.920px;"
        />
      </svg>
    </div>
  </div>
</div>
`;
